src/lines.jl

```diff
@@ -68,103 +65,118 @@ function graphline(g::AbstractGraph{T}, locs_x, locs_y, nodesize::Real) where {T
         j = dst(e)
         Δx = locs_x[j] - locs_x[i]
         Δy = locs_y[j] - locs_y[i]
-        d  = sqrt(Δx^2 + Δy^2)
-        θ  = atan(Δy,Δx)
+        θ = atan(Δy,Δx)
         startx = locs_x[i] + nodesize*cos(θ)
         starty = locs_y[i] + nodesize*sin(θ)
-        endx  = locs_x[i] + (d-nodesize)*1.00*cos(θ)
-        endy  = locs_y[i] + (d-nodesize)*1.00*sin(θ)
+        endx = locs_x[j] + nodesize*cos(θ+π)
+        endy = locs_y[j] + nodesize*sin(θ+π)
         lines[e_idx] = [(startx, starty), (endx, endy)]
     end
     return lines
 end
 
 function graphcurve(g::AbstractGraph{T}, locs_x, locs_y, nodesize::Vector{<:Real}, arrowlength, angleoffset, outangle=pi/5) where {T<:Integer}
-    lines = Array{Vector}(ne(g))
+    curves = Vector{Vector{Tuple{Float64,Float64}}}(undef, ne(g))
     arrows = Array{Vector{Tuple{Float64,Float64}}}(undef, ne(g))
     for (e_idx, e) in enumerate(edges(g))
         i = src(e)
         j = dst(e)
         Δx = locs_x[j] - locs_x[i]
         Δy = locs_y[j] - locs_y[i]
-        d  = sqrt(Δx^2 + Δy^2)
-        θ  = atan(Δy,Δx)
-        startx = locs_x[i] + nodesize[i]*cos(θ)
-        starty = locs_y[i] + nodesize[i]*sin(θ)
-        endx  = locs_x[i] + (d-nodesize[j])*1.00*cos(θ)
-        endy  = locs_y[i] + (d-nodesize[j])*1.00*sin(θ)
-        lines[e_idx] = curveedge(startx, starty, endx, endy, outangle)
-        if startx <= endx
-            arr1, arr2 = curvearrowcoords1(θ, outangle, endx, endy, arrowlength, angleoffset)
-            arrows[e_idx] = [arr1, (endx, endy), arr2]
-        else
-            arr1, arr2 = curvearrowcoords2(θ, outangle, endx, endy, arrowlength, angleoffset)
-            arrows[e_idx] = [arr1, (endx, endy), arr2]
+        θ = atan(Δy,Δx)
+        startx = locs_x[i] + nodesize[i]*cos(θ+outangle)
+        starty = locs_y[i] + nodesize[i]*sin(θ+outangle)
+        endx = locs_x[j] + nodesize[j]*cos(θ+π-outangle)
+        endy = locs_y[j] + nodesize[j]*sin(θ+π-outangle)
+
+        d = hypot(endx-startx, endy-starty)
+
+        if i == j
+            d = 2 * π * nodesize[i]
         end
+
+        curves[e_idx] = curveedge(startx, starty, endx, endy, θ, outangle, d)
+
+        arr1, arr2 = arrowcoords(θ-outangle, endx, endy, arrowlength, angleoffset)
+        arrows[e_idx] = [arr1, (endx, endy), arr2]
     end
-    return lines, arrows
+    return vcat.(curves...), arrows
```

Comment:
This kind of hurts my brain :) Maybe put a comment that on what it does (transpose vector of vectors). Although I am not yet sure, if I completely understand why we do this.

Also, consider that using the splatting operator (...) as often some performance issues, so maybe there is a better way?

Reply:
I wasn't happy with this line. It wasn't immediately obvious to me what would be an elegant way to collect then transform the data for `curve(anchor1::Vector{Tuple{<:Number, <:Number}}, control1::Vector{Tuple{<:Number, <:Number}}, control2::Vector{Tuple{<:Number, <:Number}}, anchor2::Vector{Tuple{<:Number, <:Number}})`.

Reply:
The `vcat.(curves...)` transposes the vector of vectors.

```julia
julia> c1 = [(1,2), (3,4), (5,6), (7,8)]
4-element Array{Tuple{Int64,Int64},1}:
 (1, 2)
 (3, 4)
 (5, 6)
 (7, 8)

julia> c2 = [(9,10), (11,12), (13,14), (15,16)]
4-element Array{Tuple{Int64,Int64},1}:
 (9, 10) 
 (11, 12)
 (13, 14)
 (15, 16)

julia> c = [c1, c2]
2-element Array{Array{Tuple{Int64,Int64},1},1}:
 [(1, 2), (3, 4), (5, 6), (7, 8)]       
 [(9, 10), (11, 12), (13, 14), (15, 16)]

julia> showsplat = (args...) -> show(args)
#3 (generic function with 1 method)

julia> showsplat(c...)
(Tuple{Int64,Int64}[(1, 2), (3, 4), (5, 6), (7, 8)], Tuple{Int64,Int64}[(9, 10), (11, 12), (13, 14), (15, 16)])

julia> r = vcat.(c...)
4-element Array{Array{Tuple{Int64,Int64},1},1}:
 [(1, 2), (9, 10)] 
 [(3, 4), (11, 12)]
 [(5, 6), (13, 14)]
 [(7, 8), (15, 16)]

julia> showsplat(r...) # where r... is the required form for Compose.curve.
(Tuple{Int64,Int64}[(1, 2), (9, 10)], Tuple{Int64,Int64}[(3, 4), (11, 12)], Tuple{Int64,Int64}[(5, 6), (13, 14)], Tuple{Int64,Int64}[(7, 8), (15, 16)])

```